lib/server.js

```diff
@@ -1222,7 +980,122 @@ Server.prototype._setupRequest = function _setupRequest(req, res) {
     if (self.name !== '') {
         res.setHeader('Server', self.name);
     }
-    res.version = self.router.versions[self.router.versions.length - 1];
+
+    // Request lifecycle events
+    // attach a listener for 'close' events, this will let us set
+    // a flag so that we can stop processing the request if the client closes
+    // the connection (or we lose the connection).
+    function onClose() {
+        res.removeListener('finish', onFinish);
+        res.removeListener('error', onError);
+        req._connectionState = 'close';
+
+        self._finishReqResCycle(req, res, new errors.RequestCloseError());
+    }
+    req.once('close', onClose);
+
+    // Response lifecycle events
+    function onFinish() {
+        req.removeListener('close', onClose);
+        res.removeListener('error', onError);
+        self._finishReqResCycle(req, res, res.err);
+    }
+    function onError(err) {
+        req.removeListener('close', onClose);
+        res.removeListener('finish', onFinish);
+        self._finishReqResCycle(req, res, err);
+    }
+    res.once('finish', onFinish);
+    res.once('error', onError);
+
+    // attach a listener for the response's 'redirect' event
+    res.on('redirect', function onRedirect(redirectLocation) {
+        self.emit('redirect', redirectLocation);
+    });
+};
+
+/**
+ * Wrapper method for emitting the after event. this is needed in scenarios
+ * where the async formatter has failed, and the ot assume that the
```

Comment:
This comment looks a little outdated since we don't have async formatters anymore - don't want to bikeshed on this PR but could make a note to come back and clean up comments in another PR.